graphsync.md

```diff
@@ -0,0 +1,159 @@
+  - **Selector**: Some identifier (flag) together with some data describing a traversal in the DAG
+  - **Block**: A CID together with the corresponding binary data
+
+
+General architecture
+--------------------
+
+GraphSync needs to do a lot of things in the background like verification and error handling when connecting to different peers. Hence there is a difference  between the Client-Server interface and the actual wire protocol.
+
+```
+┌───────────────────┐     ┌──────┐
+│       Local       │     │Remote│
+│                   │     │      │
+│Client <─> Consumer│ <─> │Server│
+└───────────────────┘     └──────┘
+```
+
+GraphSync (both the Consumer and the Server) returns a stream of Blocks. The order of the Blocks is the same as if it was a local traversal.
+
+The Consumer verifies the Blocks to make sure there are no malicious ones. It might also apply some filters as requested by the Client.
+
+
+### Client
+
+The Client requests a sub-DAG with a CID and a Selector and receives a stream of Blocks.
+
+### Consumer
+
+Some filtering requested by the Selector sent from the Client might not be possible without additional verification, that needs a bigger set of blocks. Hence the Consumer might send a modified Selector to the Server, which fulfills the needs in order to verify the result.
+
+After the verification and possible additional filtering of the Blocks that were returned by the Server, those Blocks are returned to the Client.
+
+So far things were simplified to a single Server. In reality the Consumer will connect to several peers. Not all of them might have all the data locally available, that is needed to fulfill the request. The Consumer deals with all the challenges related to connecting to several peers and possible errors. The Client will only receive an error if it can't be resolved by the Consumer.
+
+
+### Server
+
+The Selectors the Server understands is a subset from those the Consumer can process. A Server might only contain a subset of the data that is needed to fulfill the request. If that's the case, then an error message is returned which contains the CID of the Block that is missing as well as all Blocks that are needed for further verification.
+
+
+Peers with subsets only
+-----------------------
+
+This section describes how the Consumer is dealing with peers that only contain a subset of the data that is needed in order to fulfill the request. It is not about error handling for cases like connectivity issues.
+
+The following pseudo code describes a possible algorithm:
+
+```
+const peers = <connected-peers>
+
+const doGraphSync = function* (peers, cid, selector) {
+  const peer = peers.nextPeer()
+  const messages = peer.graphSync(cid, selector)
+
+  for (const message of messages) {
+    if (message.isBlock()) {
+      yield {
+        cid: message.cid,
+        block: message.block
+      }
+    }
+    // Server has only a subset of the requested DAG
```

Comment:
Is this saying that we are adding an "I dont have this" response message?

Reply:
That would be my plan.